Fixes Type name to delete project/screen

beinganonymous opened the pull request:
- [ ] My code follows the:
    - [x] [Google Java style guide](https://google.github.io/styleguide/javaguide.html) (for .java files)
- [x] `ant tests` passes on my machine

For all other changes:

- [x] I branched from `master`
- [x] My pull request has `master` as the base

What does this PR accomplish?

I have added ``Type name to delete screen`` feature, where user have to type accurate screen name to delete screen to avoid accidental deletion. Also this PR automatically allows user to save current checkpoint while deleting screen.

Fixes #2071  .

![app](https://user-images.githubusercontent.com/30001594/76657982-167e5080-6599-11ea-810c-124a4614d076.gif)

AppInventorWorkerBee commented:
Can one of the admins verify this patch?

pavi2410 commented:
@AppInventorWorkerBee ok to test

SusanRatiLane commented:
From testing: 

The Remove Screen probably shouldn't prepopulate with any screen name, even Screen1. 
<img width="428" alt="Screen Shot 2020-03-23 at 4 02 51 PM" src="https://user-images.githubusercontent.com/43011595/77358368-1bbd6700-6d20-11ea-8728-fa07613290f5.png">

When you click OK with a screen name that doesn't match, there is no response at all. There should probably be a notification to the user stating why the delete did not go through.

ewpatton commented:
@SusanRatiLane @beinganonymous I would suggest having a KeyUp handler that validates by coloring the textbox red or not (this is what GitHub does to delete a repo). We should also have text at the beginning to the effect of:

"Warning: You are deleting a screen. This operation cannot be undone. Please type the name of the screen ({0}) into the box below to confirm you really want to do this."

ewpatton commented:
Possibly also change the checkbox text to "Create checkpoint before deleting"

SusanRatiLane commented:
> @SusanRatiLane @beinganonymous I would suggest having a KeyUp handler that validates by coloring the textbox red or not (this is what GitHub does to delete a repo). We should also have text at the beginning to the effect of:
> 
> "Warning: You are deleting a screen. This operation cannot be undone. Please type the name of the screen ({0}) into the box below to confirm you really want to do this."

This warning currently appears when OK is clicked with a match as a final "Are you sure?" confirmation.

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +69,195 @@ public boolean willCallExecuteNextCommand() {
+            }
+          }
+        }
+      }*/
```
If this is the final version, please remove commented code.

beinganonymous replied:
Removed Commented Code

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +69,195 @@ public boolean willCallExecuteNextCommand() {
+        public void onKeyUp(KeyUpEvent event) {
+          int keyCode = event.getNativeKeyCode();
+          if (keyCode == KeyCodes.KEY_ENTER) {
+            //handleOkClick(projectRootNode);
```
Is there a reason we're checking for the enter key here? The click call is commented out.

beinganonymous replied:
Removed KeyUp Handler

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +69,195 @@ public boolean willCallExecuteNextCommand() {
+        cancelText = MESSAGES.cancelScreenButton();
+
+        // okText = "Add";
+        // cancelText = "Don't Add";
```
Commented code.

beinganonymous replied:
Removed Commented Code

ewpatton commented:
> > @SusanRatiLane @beinganonymous I would suggest having a KeyUp handler that validates by coloring the textbox red or not (this is what GitHub does to delete a repo). We should also have text at the beginning to the effect of:
> > "Warning: You are deleting a screen. This operation cannot be undone. Please type the name of the screen ({0}) into the box below to confirm you really want to do this."
> 
> This warning currently appears when OK is clicked with a match as a final "Are you sure?" confirmation.

The text I wrote included instructions for how to use the form. I'm not sure that it is obvious to most why you would need to type a screen name in. "After all, the app already knows what screen I am on." And "why do I have to type the current screen name versus the one I want to delete" (which may not be the current one).

Also, why does the textbox prepopulate with Screen1? Screen1 can never be removed. The text should be blank.

The error message that appears after you click OK is the one that appears in prod to confirm deletion. It might be worth keeping as one last failsafe, but otherwise it could probably go after this feature is merged as it is meant to be more of a confirmation than just clicking OK.

beinganonymous commented:
> @SusanRatiLane @beinganonymous I would suggest having a KeyUp handler that validates by coloring the textbox red or not (this is what GitHub does to delete a repo). We should also have text at the beginning to the effect of:
> 
> "Warning: You are deleting a screen. This operation cannot be undone. Please type the name of the screen ({0}) into the box below to confirm you really want to do this."

@ewpatton Should i add feature like if user types wrong screen name then textbox should be red or if user types right screen name then textbox should be green ?

ewpatton commented:
@beinganonymous I think red versus normal is sufficient for this. Because the operation is destructive, I don't think we want to use green at all if possible. Like I've previously said, I think how GitHub does this for repo deletion is a good model of how it should look/behave.

beinganonymous commented:
@ewpatton @SusanRatiLane I have implemented requested changes in above commits.
 - [x] The Remove Screen won't prepopulate with any screen name.
 - [x] If screen name doesn't match then notification to the user will be given. 
 - [x] Added KeyUp handler where TextBox Border turns into red color if text matches with Screen Name.
 - [x] Added Warning Message to the Dialog Box.
 - [x] Removed Commented Code.
 - [x] Changed the checkbox text to "Create checkpoint before deleting"

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java:
```diff
@@ -117,6 +117,10 @@
   @Description("Label of the button for checkpoint")
   String checkpointButton();
+  
+  @DefaultMessage("Create checkpoint before deleting"")
```
I had a compile fail probably caused by the quotes typo here.

beinganonymous replied:
Resolved the quotes typo

SusanRatiLane replied:
It looks like there are now merge conflicts in the PR.

beinganonymous commented:
Resolved all Merge Conflicts

SusanRatiLane commented:
This is what I see when I type a form name that doesn't match. The top-screen message asks the user to "select" a valid screen name, and nothing turns red.

You did show the error at the top in your video demo, but I didn't even notice it. It's not really where you think to look. If we're going to have an error message, it should say, "Please type a valid screen name," but I'm not sure we need the error message at all. The field border turning red should be enough.

<img width="457" alt="Screen Shot 2020-03-26 at 9 54 34 AM" src="https://user-images.githubusercontent.com/43011595/77654861-ff9c0e80-6f47-11ea-935d-0704cd2f7fb4.png">

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +70,158 @@ public boolean willCallExecuteNextCommand() {
+    return Window.confirm(MESSAGES.reallyDeleteFile());
+  }
+  
+  private class NewFormDialog extends DialogBox {
```
This class should have a more meaningful name. We're deleting a form, not adding one.

beinganonymous commented:
@SusanRatiLane Updated following changes
 - [x] TextBox border will turn red if screen name is wrong.
 - [x] Classname NewFormDialog changed to DeleteFormDialog.

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +70,158 @@ public boolean willCallExecuteNextCommand() {
+  		  executionFailedOrCanceled();
+  		  throw new IllegalArgumentException("node must be a YoungAndroidProjectNode");
+	  }
+  }
```
Strange whitespace here. Our style guide calls for two spaces for each of these indents.

beinganonymous replied:
Removed whitespace

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +70,158 @@ public boolean willCallExecuteNextCommand() {
+  private class DeleteFormDialog extends DialogBox {
+    // UI elements
+    private final LabeledTextBox newNameTextBox;
+	private final CheckBox cb = new CheckBox(MESSAGES.checkBoxButton());
```
Extra space here.

beinganonymous replied:
Removed extra space

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -36,6 +70,158 @@ public boolean willCallExecuteNextCommand() {
+        }
+      });
+   
+      if (projectRootNode instanceof YoungAndroidFormNode) {
```
Whitespace here needs to be two spaces per indent.

beinganonymous replied:
Added two whitespace per indent

SusanRatiLane commented on appinventor/appengine/src/com/google/appinventor/client/explorer/commands/DeleteFileCommand.java:
```diff
@@ -144,40 +143,38 @@ public void onClick(ClickEvent event) {
+  	labelPanel.setSize("100%", "22px");
+  	contentPanel.add(labelPanel);
+  	contentPanel.add(newNameTextBox);
+        HorizontalPanel buttonPanel = new HorizontalPanel();
```
Hmm. The whitespace here is still weird. Are you using tab key for spacing? You're probably better off using spaces instead.

beinganonymous replied:
Sorry actually i was using tab key instead of spaces, i will keep it in the mind to use spaces instead.
Removed all weird whitespaces
